# Avoid in-place zip deletion for Bazel's Netty tcnative jar. On some systems
# Info-ZIP exits with "Interrupted (aborting)" while deleting from this jar, so
# rebuild the filtered jar from extracted contents with the JDK's jar tool
# instead. Avoid Info-ZIP entirely, since it can also be interrupted while
# creating the replacement archive on lock-sensitive filesystems.
diff --git a/third_party/BUILD b/third_party/BUILD
index 1f35568..c8f1b35 100644
--- a/third_party/BUILD
+++ b/third_party/BUILD
@@ -500,16 +500,16 @@ alias(
 )
 
 UNNECESSARY_DYNAMIC_LIBRARIES = select({
-    "//src/conditions:windows": "*.so *.jnilib",
-    "//src/conditions:darwin": "*.so *.dll",
-    "//src/conditions:linux_x86_64": "*.jnilib *.dll",
-    "//src/conditions:linux_s390x": "*.jnilib *.dll",
+    "//src/conditions:windows": "find $$NETTY_JAR_TMPDIR -name '*.so' -delete && find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && ",
+    "//src/conditions:darwin": "find $$NETTY_JAR_TMPDIR -name '*.so' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
+    "//src/conditions:linux_x86_64": "find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
+    "//src/conditions:linux_s390x": "find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
     # The .so file is an x86/s390x one, so we can just remove it if the CPU is not x86/s390x
-    "//src/conditions:arm": "*.so *.jnilib *.dll",
-    "//src/conditions:linux_aarch64": "*.so *.jnilib *.dll",
-    "//src/conditions:linux_ppc": "*.so *.jnilib *.dll",
-    "//src/conditions:freebsd": "*.so *.jnilib *.dll",
-    "//src/conditions:openbsd": "*.so *.jnilib *.dll",
+    "//src/conditions:arm": "find $$NETTY_JAR_TMPDIR -name '*.so' -delete && find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
+    "//src/conditions:linux_aarch64": "find $$NETTY_JAR_TMPDIR -name '*.so' -delete && find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
+    "//src/conditions:linux_ppc": "find $$NETTY_JAR_TMPDIR -name '*.so' -delete && find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
+    "//src/conditions:freebsd": "find $$NETTY_JAR_TMPDIR -name '*.so' -delete && find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
+    "//src/conditions:openbsd": "find $$NETTY_JAR_TMPDIR -name '*.so' -delete && find $$NETTY_JAR_TMPDIR -name '*.jnilib' -delete && find $$NETTY_JAR_TMPDIR -name '*.dll' -delete && ",
     # Default is to play it safe -- better have a big binary than a slow binary
     #   The empty string means nothing is to be removed from the library;
     #   the rule command tests for the empty string explictly to avoid
@@ -530,10 +530,15 @@ genrule(
         "//conditions:default": [],
     }),
     outs = ["netty_tcnative/netty-tcnative-filtered.jar"],
-    cmd = "cp $< $@ && " +
-          # Make sure we can write the output file, even if the input isn't writable.
-          "chmod +w $@ && " +
-          "zip -qd $@ */license/* " + UNNECESSARY_DYNAMIC_LIBRARIES,
+    cmd = "mkdir -p $$(dirname $@) && " +
+          "OUTPUT=$$(pwd)/$@ && " +
+          "rm -f $$OUTPUT && " +
+          "NETTY_JAR_TMPDIR=$$(mktemp -d) && " +
+          "trap 'chmod -R +w $$NETTY_JAR_TMPDIR && rm -rf $$NETTY_JAR_TMPDIR' EXIT && " +
+          "unzip -q $< -d $$NETTY_JAR_TMPDIR && " +
+          "find $$NETTY_JAR_TMPDIR -path '*/license/*' -delete && " +
+          UNNECESSARY_DYNAMIC_LIBRARIES +
+          "jar --create --no-manifest --file $$OUTPUT -C $$NETTY_JAR_TMPDIR .",
 )
 
 distrib_java_import(
